include/gmsh.tcc

```diff
@@ -4,137 +4,151 @@
+    read_elements(file);
+  }
+  file.close();
+}
+
+//! Read keyword
+//! \tparam Tdim Dimension
+//! \tparam Tvertices Number of vertices in element
+//! \param[in] filename Input mesh filename
+//! \param[in] keyword search for the word of interest
+template <unsigned Tdim, unsigned Tvertices>
+void GMSH<Tdim, Tvertices>::read_keyword(std::ifstream& file,
+                                         const std::string& keyword) {
+
+  bool read_status = false;
+  std::string line;
+  file.clear();
+  file.seekg(0, std::ios::beg);
+  while (std::getline(file, line)) {
+    if (line != keyword) {
+      if (line.find(keyword) != std::string::npos) {
+        break;
+      };
+    } else {
+      std::cout << "Read keyword: " << keyword << " successfully\n";
+      read_status = true;
+      break;
+    }
+  }
+  if (!read_status) {
+    std::cerr << "Cannot find keyword: " << keyword << '\n';
+  }
 }
 
 //! Read vertex id and coordinates in GMSH
 //! \tparam Tdim Dimension
 //! \tparam Tvertices Number of vertices in element
 //! \param[in] filename Input mesh filename
 template <unsigned Tdim, unsigned Tvertices>
-void GMSH<Tdim, Tvertices>::read_vertices(const std::string& filename) {
+void GMSH<Tdim, Tvertices>::read_vertices(std::ifstream& file) {
+
+  //! Find the line of interest
+  read_keyword(file, "$Nodes");
 
-  //! Number of vertices
-  double nvertices = std::numeric_limits<double>::max();
-  const unsigned toplines = 4;
+  std::string line;
+  std::getline(file, line);
+  std::istringstream istream(line);
+
+  //! Read number of vertices
+  unsigned nvertices = std::numeric_limits<unsigned>::max();
+  istream >> nvertices;
+  getline(istream, line);
 
   //! Vertices id
   unsigned vertid = 0;
 
-  std::fstream infile;
-  infile.open(filename, std::ios::in);
-
-  //! Check if input file is good
-  if (infile.good()) {
-    std::cout << "Vertices file found" << '\n';
+  //! Read vertex coordinates & id
+  for (unsigned i = 0; i < nvertices; ++i) {
+    std::getline(file, line);
+    std::istringstream istream(line);
 
-    std::string line;
+    if (line.find('#') == std::string::npos && line != "") {
+      //! Coordinates of vertex
+      std::array<double, Tdim> vertex;
 
-    //! Ignore first 4 lines
-    for (unsigned i = 0; i < toplines; ++i) {
-      std::getline(infile, line);
-    }
-    //! Read number of vertices
-    infile >> nvertices;
-    getline(infile, line);
-
-    //! Read vertex coordinates & id
-    for (unsigned i = 0; i < nvertices; ++i) {
-      std::getline(infile, line);
-      std::istringstream istream(line);
-
-      if (line.find('#') == std::string::npos && line != "") {
-        //! Coordinates of vertex
-        std::array<double, Tdim> vertex;
-
-        istream >> vertid;
-        istream >> vertex.at(0) >> vertex.at(1) >> vertex.at(2);
-        vertices_.insert(std::make_pair(vertid, vertex));
-      }
+      istream >> vertid;
+      istream >> vertex.at(0) >> vertex.at(1) >> vertex.at(2);
```

Comment:
Is this true for 2D and 3D?

Reply:
in 2D, we will have the z coordinate as 0 right? so this is a general statement, no?

Reply:
Is that true for GMSH?